controllers/clustertopology_controller_compute.go

```diff
@@ -140,6 +141,149 @@ func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clust
+	if err != nil {
+		return nil, errors.Wrap(err, "cluster infrastructure state")
+	}
+	infra, err := external.Get(ctx, r.Client, cluster.Spec.InfrastructureRef, cluster.Namespace)
+	if err != nil {
+		return nil, errors.Wrap(err, "cluster infrastructure state")
+	}
+	return infra, nil
+}
+
+// getControlPlaneState returns information on the control plane being used by the cluster. If there is no reference
+// no error is thrown and reconciliation continues. If a reference is set but not found, through an error or not existing
+// in the api server an error is thrown. If the control plane requires infrastructure as configured in the clusterClass
+// an error will be thrown if the control plane is found but with no infrastructure.
+func (r *ClusterTopologyReconciler) getControlPlaneState(ctx context.Context, cluster *clusterv1.Cluster, class clusterv1.ClusterClass) (controlPlaneTopologyState, error) {
+	err := utilconversion.ConvertReferenceAPIContract(ctx, r.Client, r.restConfig, cluster.Spec.ControlPlaneRef)
+	if err != nil {
+		return controlPlaneTopologyState{}, errors.Wrap(err, "control plane")
+	}
+	cp, err := external.Get(ctx, r.Client, cluster.Spec.ControlPlaneRef, cluster.Namespace)
+	if err != nil {
+		return controlPlaneTopologyState{}, errors.Wrap(err, "control plane")
+	}
+	// Some control plane providers may not require infrastructure to run. This check returns early if the field
+	// indicating infrastructure is required is not found in the clusterClass of the given cluster
+	if class.Spec.ControlPlane.MachineInfrastructure == nil {
+		return controlPlaneTopologyState{
+			cp,
+			nil,
+		}, nil
+	}
+	cpInfraRef, err := getNestedRef(cp, "spec", "machineTemplate", "infrastructureRef")
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrap(err, "control plane infra ref")
+	}
+	err = utilconversion.ConvertReferenceAPIContract(ctx, r.Client, r.restConfig, cpInfraRef)
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrap(err, "control plane infrastructure reference not found")
+	}
+	infra, err := external.Get(ctx, r.UnstructuredCachingClient, cpInfraRef, cpInfraRef.Namespace)
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrap(err, "control plane infrastructure reference not found")
+	}
+	return controlPlaneTopologyState{
+		cp,
+		infra,
+	}, nil
+}
+
+// getCurrentMachineDeploymentState queries the apiserver for all MachineDeployments linked to a given cluster and managed
+// through a ClusterClass using labels. If it finds MachineDeployments it inspects them for infrastructure and bootstrap
+// references. Where these are not found the function will throw an error.
+func (r *ClusterTopologyReconciler) getCurrentMachineDeploymentState(ctx context.Context, cluster *clusterv1.Cluster) ([]machineDeploymentTopologyState, error) {
+	mDeploymentState := make([]machineDeploymentTopologyState, 0)
+
+	ms := &clusterv1.MachineDeploymentList{}
```

Comment:
nit:
```suggestion
	md := &clusterv1.MachineDeploymentList{}
```
ms is usually used for MachineSets